File: sdk/python/mashgate/resources/kyc.py
"""Tenant-scoped KYC capability."""

from __future__ import annotations

from enum import Enum
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from mashgate.client import MashgateClient


class KycStatus(str, Enum):
    UNSPECIFIED = "KYC_STATUS_UNSPECIFIED"
    PENDING = "KYC_STATUS_PENDING"
    IN_REVIEW = "KYC_STATUS_IN_REVIEW"
    PASSED = "KYC_STATUS_PASSED"
    FAILED = "KYC_STATUS_FAILED"
    EXPIRED = "KYC_STATUS_EXPIRED"
    OVERRIDDEN = "KYC_STATUS_OVERRIDDEN"


class KycSubjectType(str, Enum):
    INDIVIDUAL = "KYC_SUBJECT_INDIVIDUAL"
    BUSINESS = "KYC_SUBJECT_BUSINESS"


class KycCheckType(str, Enum):
    IDENTITY = "KYC_CHECK_IDENTITY"
    AML = "KYC_CHECK_AML"
    SANCTIONS = "KYC_CHECK_SANCTIONS"
    PEP = "KYC_CHECK_PEP"
    FULL = "KYC_CHECK_FULL"


class KYCResource:
    def __init__(self, client: MashgateClient) -> None:
        self._c = client

    def request_check(
        self,
        *,
        subject_id: str,
        subject_type: KycSubjectType | str,
        check_type: KycCheckType | str,
        idempotency_key: str,
        provider: str | None = None,
        metadata: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "tenantId": self._c.require_tenant_id(),
            "subjectId": subject_id,
            "subjectType": subject_type,
            "checkType": check_type,
        }
        if provider is not None:
            body["provider"] = provider
        if metadata is not None:
            body["metadata"] = metadata
        return self._c.request(
            "POST",
            "/v1/kyc/checks",
            body=body,
            extra_headers={"Idempotency-Key": idempotency_key},
        )

    def get(self, check_id: str) -> dict[str, Any]:
        return self._c.request(
            "GET",
            f"/v1/kyc/checks/{check_id}",
            query={"tenant_id": self._c.require_tenant_id()},
        )

    def list(
        self,
        *,
        subject_id: str | None = None,
        status: KycStatus | str | None = None,
        limit: int | None = None,
        cursor: str | None = None,
    ) -> dict[str, Any]:
        return self._c.request(
            "GET",
            "/v1/kyc/checks",
            query={
                "tenant_id": self._c.require_tenant_id(),
                "subject_id": subject_id,
                "status": status,
                "limit": limit,
                "cursor": cursor,
            },
        )

    def override(
        self,
        check_id: str,
        *,
        status: KycStatus | str,
        override_note: str,
    ) -> dict[str, Any]:
        return self._c.request(
            "POST",
            f"/v1/kyc/checks/{check_id}/override",
            body={
                "tenantId": self._c.require_tenant_id(),
                "checkId": check_id,
                "status": status,
                "overrideNote": override_note,
            },
        )

File: sdk/typescript/src/resources/kyc.ts
import type { MashgateClient } from "../client.js";

export const KycStatus = {
  Unspecified: "KYC_STATUS_UNSPECIFIED",
  Pending: "KYC_STATUS_PENDING",
  InReview: "KYC_STATUS_IN_REVIEW",
  Passed: "KYC_STATUS_PASSED",
  Failed: "KYC_STATUS_FAILED",
  Expired: "KYC_STATUS_EXPIRED",
  Overridden: "KYC_STATUS_OVERRIDDEN",
} as const;
export type KycStatus = (typeof KycStatus)[keyof typeof KycStatus];

export const KycSubjectType = {
  Individual: "KYC_SUBJECT_INDIVIDUAL",
  Business: "KYC_SUBJECT_BUSINESS",
} as const;
export type KycSubjectType = (typeof KycSubjectType)[keyof typeof KycSubjectType];

export const KycCheckType = {
  Identity: "KYC_CHECK_IDENTITY",
  AML: "KYC_CHECK_AML",
  Sanctions: "KYC_CHECK_SANCTIONS",
  PEP: "KYC_CHECK_PEP",
  Full: "KYC_CHECK_FULL",
} as const;
export type KycCheckType = (typeof KycCheckType)[keyof typeof KycCheckType];

export interface KycRiskSignal {
  code: string;
  description: string;
  severity: string;
}

export interface KycCheck {
  checkId: string;
  tenantId: string;
  subjectId: string;
  subjectType: KycSubjectType;
  checkType: KycCheckType;
  status: KycStatus;
  provider: string;
  providerRef: string;
  riskSignals: KycRiskSignal[];
  failureCode: string;
  failureReason: string;
  overrideBy: string;
  overrideNote: string;
  createdAt: string;
  updatedAt: string;
  expiresAt?: string;
}

export interface RequestKycCheck {
  subjectId: string;
  subjectType: KycSubjectType;
  checkType: KycCheckType;
  provider?: string;
  metadata?: Record<string, string>;
}

export interface RequestKycCheckResponse {
  check: KycCheck;
  redirectUrl?: string;
}

export interface ListKycChecksQuery {
  subjectId?: string;
  status?: KycStatus;
  limit?: number;
  cursor?: string;
}

export interface ListKycChecksResponse {
  checks: KycCheck[];
  nextCursor?: string;
}

export class KycResource {
  constructor(private readonly client: MashgateClient) {}

  request(data: RequestKycCheck, idempotencyKey: string): Promise<RequestKycCheckResponse> {
    return this.client.request("POST", "/v1/kyc/checks", {
      body: { ...data, tenantId: this.client.requireTenantId() },
      headers: { "Idempotency-Key": idempotencyKey },
    });
  }

  get(checkId: string): Promise<KycCheck> {
    return this.client.request("GET", `/v1/kyc/checks/${checkId}`, {
      query: { tenant_id: this.client.requireTenantId() },
    });
  }

  list(query: ListKycChecksQuery = {}): Promise<ListKycChecksResponse> {
    return this.client.request("GET", "/v1/kyc/checks", {
      query: {
        tenant_id: this.client.requireTenantId(),
        subject_id: query.subjectId,
        status: query.status,
        limit: query.limit,
        cursor: query.cursor,
      },
    });
  }

  override(checkId: string, status: KycStatus, overrideNote: string): Promise<KycCheck> {
    return this.client.request("POST", `/v1/kyc/checks/${checkId}/override`, {
      body: {
        tenantId: this.client.requireTenantId(),
        checkId,
        status,
        overrideNote,
      },
    });
  }
}
